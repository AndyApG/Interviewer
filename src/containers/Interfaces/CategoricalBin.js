import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withStateHandlers } from 'recompose';
import PropTypes from 'prop-types';
import withPrompt from '../../behaviours/withPrompt';
import PromptSwiper from '../PromptSwiper';
import CategoricalList from '../CategoricalList';
import { makeGetPromptVariable, getPromptOtherVariable, makeNetworkNodesForType } from '../../selectors/interface';
import MultiNodeBucket from '../../components/MultiNodeBucket';
import { entityAttributesProperty } from '../../ducks/modules/network';

const categoricalBinStateHandler = withStateHandlers(
  {
    expandedBinIndex: null,
  },
  {
    handleExpandBin: () =>
      (expandedBinIndex = null) => ({ expandedBinIndex }),
  },
);

/**
  * CategoricalBin Interface
  */
const CategoricalBin = ({
  promptForward,
  promptBackward,
  prompt,
  uncategorizedNodes,
  expandedBinIndex,
  handleExpandBin,
  stage,
}) => {
  const {
    prompts,
  } = stage;
  return (
    <div className="categorical-bin-interface">
      <div className="categorical-bin-interface__prompt">
        <PromptSwiper
          forward={promptForward}
          backward={promptBackward}
          prompt={prompt}
          prompts={prompts}
        />
      </div>
      <div className="categorical-bin-interface__bucket" onClick={() => handleExpandBin()}>
        <MultiNodeBucket
          nodes={uncategorizedNodes}
          listId={`${stage.id}_${prompt.id}_CAT_BUCKET`}
          sortOrder={prompt.bucketSortOrder}
        />
      </div>
      <CategoricalList
        key={prompt.id}
        stage={stage}
        prompt={prompt}
        expandedBinIndex={expandedBinIndex}
        onExpandBin={handleExpandBin}
      />
    </div>
  );
};

CategoricalBin.propTypes = {
  stage: PropTypes.object.isRequired,
  prompt: PropTypes.object.isRequired,
  promptForward: PropTypes.func.isRequired,
  promptBackward: PropTypes.func.isRequired,
  uncategorizedNodes: PropTypes.array.isRequired,
  expandedBinIndex: PropTypes.string.isRequired,
  handleExpandBin: PropTypes.func.isRequired,
};

function makeMapStateToProps() {
  const getStageNodes = makeNetworkNodesForType();
  const getPromptVariable = makeGetPromptVariable();

  return function mapStateToProps(state, props) {
    const stageNodes = getStageNodes(state, props);
    const activePromptVariable = getPromptVariable(state, props);
    const [promptOtherVariable] = getPromptOtherVariable(state, props);

    const matchNoCategory = node =>
      !node[entityAttributesProperty][activePromptVariable] &&
      !node[entityAttributesProperty][promptOtherVariable];

    return {
      activePromptVariable,
      uncategorizedNodes: stageNodes.filter(matchNoCategory),
    };
  };
}

export { CategoricalBin as UnconnectedCategoricalBin };

export default compose(
  withPrompt,
  connect(makeMapStateToProps),
  categoricalBinStateHandler,
)(CategoricalBin);
