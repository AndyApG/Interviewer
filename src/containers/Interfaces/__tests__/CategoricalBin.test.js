/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import { UnconnectedCategoricalBin as CategoricalBin } from '../CategoricalBin';

const requiredProps = {
  nodesForPrompt: [],
  prompt: {},
  stage: {},
  promptBackward: jest.fn(),
  promptForward: jest.fn(),
};

describe('CategoricalBin', () => {
  it('renders CategoricalBin interface', () => {
    const component = shallow(<CategoricalBin {...requiredProps} />);
    expect(component.find('.categorical-bin-interface')).toHaveLength(1);
    expect(component.find('.categorical-bin-interface').children()).toHaveLength(3);
    expect(component.find('.categorical-bin-interface__prompt')).toHaveLength(1);
    expect(component.find('.categorical-bin-interface__bucket')).toHaveLength(1);
  });
});
