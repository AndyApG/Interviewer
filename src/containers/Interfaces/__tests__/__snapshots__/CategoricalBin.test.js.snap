// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoricalBin renders CategoricalBin interface 1`] = `
<div
  className="categorical-bin-interface"
>
  <div
    className="categorical-bin-interface__prompt"
  >
    <Connect(PromptSwiper)
      backward={[MockFunction]}
      forward={[MockFunction]}
      prompt={Object {}}
    />
  </div>
  <div
    className="categorical-bin-interface__bucket"
    onClick={[Function]}
  >
    <MultiNodeBucket
      itemType="NODE"
      label={[Function]}
      listId="undefined_undefined_CAT_BUCKET"
      nodeColor=""
      nodes={Array []}
      sortOrder={Array []}
    />
  </div>
  <Connect(Monitor)
    prompt={Object {}}
    stage={Object {}}
  />
</div>
`;
